mirascope/core/base/_utils/_protocols.py

```diff
@@ -108,12 +150,87 @@ def __call__(
 
 
 def fn_is_async(
-    fn: Callable[_P, _R] | Callable[_P, Awaitable[_R]],
-) -> TypeIs[Callable[_P, Awaitable[_R]]]:
+    fn: Callable[_P, _R] | Callable[_P, Awaitable[_R] | Coroutine[Any, Any, _R]],
+) -> TypeIs[Callable[_P, Awaitable[_R] | Coroutine[Any, Any, _R]]]:
     return inspect.iscoroutinefunction(fn)
 
 
 class SetupCall(
+    Protocol[
+        _SyncBaseClientT,
+        _AsyncBaseClientT,
+        _BaseDynamicConfigT,
+        _BaseCallParamsT,
+        _ResponseT,
+        _ResponseChunkT,
+        _BaseToolT,
+    ]
+):
+    @overload
+    def __call__(
+        self,
+        *,
+        model: str,
+        client: _AsyncBaseClientT | None,
+        fn: Callable[..., Awaitable[_BaseDynamicConfigT]],
+        fn_args: dict[str, Any],
+        dynamic_config: _BaseDynamicConfigT,
+        tools: list[type[BaseTool] | Callable] | None,
+        json_mode: bool,
+        call_params: _BaseCallParamsT,
+        extract: bool,
+    ) -> tuple[
+        AsyncCreateFn[_ResponseT, _ResponseChunkT],
+        str | None,
+        list[Any],
+        list[type[_BaseToolT]] | None,
+        BaseCallKwargs,
+    ]: ...
+
+    @overload
+    def __call__(
+        self,
+        *,
+        model: str,
+        client: _SyncBaseClientT | None,
+        fn: Callable[..., _BaseDynamicConfigT],
+        fn_args: dict[str, Any],
+        dynamic_config: _BaseDynamicConfigT,
+        tools: list[type[BaseTool] | Callable] | None,
+        json_mode: bool,
+        call_params: _BaseCallParamsT,
+        extract: bool,
+    ) -> tuple[
+        CreateFn[_ResponseT, _ResponseChunkT],
+        str | None,
+        list[Any],
+        list[type[_BaseToolT]] | None,
+        BaseCallKwargs,
+    ]: ...
+
+    def __call__(
+        self,
+        *,
+        model: str,
+        client: _SyncBaseClientT | _AsyncBaseClientT | None,
+        fn: Callable[..., _BaseDynamicConfigT | Awaitable[_BaseDynamicConfigT]],
+        fn_args: dict[str, Any],
+        dynamic_config: _BaseDynamicConfigT,
+        tools: list[type[BaseTool] | Callable] | None,
+        json_mode: bool,
+        call_params: _BaseCallParamsT,
+        extract: bool,
+    ) -> tuple[
+        CreateFn[_ResponseT, _ResponseChunkT]
+        | AsyncCreateFn[_ResponseT, _ResponseChunkT],
+        str | None,
+        list[Any],
+        list[type[_BaseToolT]] | None,
+        BaseCallKwargs,
+    ]: ...  # pragma: no cover
+
+
+class BaseClientSetupCall(
```

Comment:
we should name this so it's reason for existing is more clear

`SameSyncAndAsyncClientSetupCall` to it's more clear this is for providers who have a client that is shared across sync + async?